specs/phase0/p2p-interface.md

```diff
@@ -243,11 +243,21 @@ Each gossipsub [message](https://github.com/libp2p/go-libp2p-pubsub/blob/master/
 Clients MUST reject (fail validation) messages that are over this size limit.
 Likewise, clients MUST NOT emit or propagate messages larger than this limit.
 
-The `message-id` of a gossipsub message MUST be the first 8 bytes of the SHA-256 hash of the message data, i.e.:
-
-```python
-   message-id: SHA256(message.data)[0:8]
-```
+The `message-id` of a gossipsub message MUST be the following 20 byte value computed from the message data:
+* If `message.data` can be snappy decompressed:
+    * Let `message-id` be the first 20 bytes of the `SHA256` hash of the snappy decompressed message data,
+      i.e. `SHA256(snappy_decompress(message.data))[:20]`.
+    * Then set the most significant bit of the left-most byte of `message-id` to 1,
+      i.e. `message-id[0] |= (0x01 << 7)`.
+* Otherwise:
```

Comment:
what is the reason we handle snappy decompressed messages differently vs normal messages ? 
In the latter's case, we do this instead 
```
   * Then set the most significant bit of the left-most byte of `message-id` to 0,
      i.e. `message-id[0] &= (0xFF >> 1)`.
```

Reply:
there are two types of messages for our purposes: Ones that can decompress and ones that cannot. Because a noncompressed messages can have multiple valid compressions, we can't use compressed messages for the input to id's (otherwise becomes a DoS vector). 

That still leaves us with messages that can't decompress validly at all. These need a message-id in gossipsub to be able to use the gossipsub spec/caches. In this case the message-id must be derived from the message.data itself.

To isolate the message-id domain from each, we set a flag bit (1 or 0 respectively)

Without that isolation, the problem  is that if you take data `D` (that is _not_ a valid input to `snappy_decompress`) and compress it to `C`, then setting `message.data` to `C` would result in the same `message-id` as if you set it to `D`.

The bit fiddling is a bit annoying. I might instead do an isolation via adding a domain to the hash inputs